### ISSUE TITLE
possible HTTP load problem

### ISSUE THREAD

**Issue by rkuhn** [akka/akka#17395]
reported on twitter by Niko Will (@win1imb), reproducer is at https://github.com/win1imb/akka-poc

**Comment by win1imb** [comment]
Hi and thanks for opening this issue. To reproduce the problem it should be enough to run the SearchService class and the corresponding Gatling test. On my machine (Windows 7, 64 Bit, 16GB RAM) around 40% of the requests end up in an error message saying

```
java.net.ConnectException: Connection refused: no further information: localhost/127.0.0.1:8080
```

**Comment by rkuhn** [comment]
Would you mind using the `akka.http.javadsl.Http` extension directly and specifying a larger value for the backlog parameter? This one defaults to 100 and if connections cannot be accepted quickly enough this will give you the observed result.

**Comment by jrudolph** [comment]
@rkuhn that's also a good reminder to add the missing Java overloads for specifying settings.

@win1imb The result of such a load test will very much depend on benchmarking tools and setup.

"10000 concurrent users" can mean a lot of different things. E.g. if you setup `wrk` to use 10000 connections, it will try to create all those (persistent HTTP) connections but at the same time it already starts hammering the server with requests. Normally `wrk` will be able to completely max out server CPU by sending just on 8 connections (or at least as many as CPUs are available) if handling the request is completely CPU-bound which is the usual benchmarking case. 

Increasing the number of connections from this point will only mean that connection establishment and request handling will battle for resources which means that timeouts are likely.

So, in summary, try to separate testing for performance (How many RPS are possible?) and scalability over connections (What is the performance impact of running requests over more connections?). From what I've seen scalability is ok while performance still needs to be improved.

/cc @sirthias

**Comment by win1imb** [comment]
@rkuhn I had a quick look but I'm familiar with the internals of akka. If I've understood it correctly, I have to use the Http.get(system).bind(....) method, right? The one where I can specify the backlog size as parameter has a lot of other parameters I have no clue about, what to pass here.

@jrudolph you're right, that's why I provided the stress test setup with gatling in my example project, too. As you said, increasing the number of connections will end up in a battle for resources. But if you have to choose a technology, one can cope with that and answer the requests even if not very fast and the other refuses the connections because it's not able to manage the battle for resources, which one would you use? In a world with SLAs and zero downtime, answering a request with 150ms (which was the mean of spring boot in my comparison) compared to 40% connection refused, I would choose the first option ;)

Don't get me wrong, I like akka and actor based programming, especially what it does to handle concurrency and resilience in a very concise way. As I said, possibly the error sits in front of the computer. But at the end of the day, we have to be confident with the technology we use.

**Comment by n1ko-w1ll** [comment]
BTW, had to change my username, so the code examples are now at https://github.com/n1ko-w1ll/akka-poc

**Comment by rkuhn** [comment]
My plan was to look into this today, but might not be able to, depending on how the traveling goes. I’ll definitely try out your project to see what’s wrong.

**Comment by jrudolph** [comment]
@n1ko-w1ll

> But if you have to choose a technology, one can cope with that and answer the requests even if not very fast and the other refuses the connections because it's not able to manage the battle for resources, which one would you use?

This assumes that the problem is related to latency (i.e. the processing pipeline is too long but the server has still some CPU capacity left). If you assume the problem is throughput related (i.e. the server just cannot keep up with work) then the result is not slow responses but a DOS scenario where response times will increase and eventually the server won't answer any more at all. If the problem is throughput-related then in spirit of reactivity you would actually choose the server that backpressures new connections (e.g. by letting connection attempts timeout and rely on them reconnecting with a backoff strategy) instead of a server which lets itself get overwhelmed with connections/requests.

What I meant before with "From what I've seen scalability is ok while performance still needs to be improved." is that akka-http may currently be just too slow to run your example but it doesn't really depend on the number of connections. So, I suspect that all that you are seeing is a consequence of akka-http being too slow.

That said, I agree with you that akka-http should be at least in the same ballpark performance-wise as spring. :)

Btw. thanks for including the gatling script and sorry that I overlooked it before.

**Comment by n1ko-w1ll** [comment]
@jrudolph okay, I agree with that. But I don't think akka-http is too slow. If I reduce the number of concurrent users to 1000 in the given test, akka-http is even a bit faster than spring.

**Comment by drewhk** [comment]
@n1ko-w1ll well, it *is* slow :) We first want to make it stable with the minimal set of features to be usable, but we will need to optimize after. I hope we get there soon!

**Comment by n1ko-w1ll** [comment]
@drewhk but still faster than spring ;)

**Comment by jrudolph** [comment]
@n1ko-w1ll As @rkuhn suggested Increasing the backlog will then probably help, but unfortunately it is next to impossible to do it in Java code right now. If you can create a Scala class in your project you could create your own version of `HttpApp` where you override the `handleConnectionsWithRoute` method with a copy that sets the backlog parameter to a higher value:

https://github.com/akka/akka/blob/release-2.3-dev/akka-http-java/src/main/scala/akka/http/javadsl/server/HttpService.scala#L34

**Comment by n1ko-w1ll** [comment]
@jrudolph that could be an option. Have to see when I can find time to test it.

**Comment by hepin1989** [comment]
@jrudolph does the io for akka-http plug-able?

**Comment by jrudolph** [comment]
@hepin1989 There is a certain kind of plugability that allows you to use the HTTP layers without TCP, if that's what you mean. This mechanism is (will be) used to support HTTPS. You can even use a transport layer that is implemented with other implementations of reactive-streams. The implementation itself depends on akka-stream. What kind of application do you have in mind?

**Comment by hepin1989** [comment]
@jrudolph that really what I have in mind.I am thinking how about using https://github.com/netty/netty-tcnative for ssl and netty/aio for tcp but using akka-http for the programming.

so all left to do is a reactive-streams around this and then could connect them all together right?
that's really great!.

**Comment by jrudolph** [comment]
@hepin1989 in theory, it could work and I'd be interested about the results. From our experience, "just a reactive-streams API around this" will still be lots of work and optimizing performance while mixing several stacks together will probably be even harder than trying to optimize just the single stack...

**Comment by hepin1989** [comment]
@jrudolph yes that's it is,so currently we are building our game server plus gm around akka and spray.
will move on akka-http later.the first thing is make it up and run,and then optimize.:) thanks for your works.

**Comment by rkuhn** [comment]
Currently it is extremely ugly to set the backlog parameter from Java, but I have hacked it together [here](https://github.com/rkuhn/akka-poc/blob/master/src/main/java/org/example/SearchService.java#L42). Running this will not succeed unless setting a few kernel parameters:
~~~
sudo sysctl -w kern.ipc.somaxconn=12000
sudo sysctl -w kern.maxfilesperproc=1048576
sudo sysctl -w kern.maxfiles=1148576
~~~
When doing so, the test result is quite different:
~~~
================================================================================
---- Global Information --------------------------------------------------------
> request count                                     298274 (OK=292862 KO=5412  )
> min response time                                      0 (OK=0      KO=60002 )
> max response time                                  60032 (OK=59997  KO=60032 )
> mean response time                                  1518 (OK=437    KO=60005 )
> std deviation                                       8843 (OK=3908   KO=2     )
> response time 50th percentile                         20 (OK=18     KO=60005 )
> response time 75th percentile                         38 (OK=36     KO=60007 )
> mean requests/sec                                3723.584 (OK=3656.022 KO=67.562)
---- Response Time Distribution ------------------------------------------------
> t < 800 ms                                        287834 ( 96%)
> 800 ms < t < 1200 ms                                 506 (  0%)
> t > 1200 ms                                         4522 (  2%)
> failed                                              5412 (  2%)
---- Errors --------------------------------------------------------------------
> java.util.concurrent.TimeoutException: Request timed out to lo   5412 (100,0%)
calhost/127.0.0.1:8080 of 60000 ms
================================================================================
~~~
What we see here is that some requests time out because the server is being overloaded, and we possibly don’t handle all non-nominal processing conditions with perfect grace yet, but it is quite clear that Akka HTTP can handle quite some load already—we will definitely improve, though! Johannes, it might be interesting how only some of the requests get “stuck”, there is a clear latency gap between almost everything going nice and fast and some requests seem to starve.

I don’t have the Spring comparison data, so it would be great if Niko could comment on these findings.

**Comment by hepin1989** [comment]
@rkuhn maybe we could add some doc about the server side env setup. like ```maxfiles```.

**Comment by n1ko-w1ll** [comment]
Hm, that does not look very impressive yet. I mean, I don't know your hardware and mine is quite powerful for a laptop (i7-4800MQ @ 2.70 GHz with 16 GB RAM). The results for spring on my machine are quite similar and my spring application already implements a query parser with parboiled, translates the query to mongo criterias and retrieves the results from MongoDB.

```
================================================================================
---- Global Information --------------------------------------------------------
> request count                                     289354 (OK=289354 KO=0     )
> min response time                                      0 (OK=0      KO=-     )
> max response time                                   1877 (OK=1877   KO=-     )
> mean response time                                    22 (OK=22     KO=-     )
> std deviation                                         12 (OK=12     KO=-     )
> response time 50th percentile                         21 (OK=21     KO=-     )
> response time 75th percentile                         28 (OK=28     KO=-     )
> mean requests/sec                                3534.526 (OK=3534.526 KO=-     )
---- Response Time Distribution ------------------------------------------------
> t < 800 ms                                        289353 (100%)
> 800 ms < t < 1200 ms                                   0 (  0%)
> t > 1200 ms                                            1 (  0%)
> failed                                                 0 (  0%)
================================================================================
```

**Comment by n1ko-w1ll** [comment]
By the way... the application will be deployed on cloudfoundry. I'm not sure if we can edit these environment variables there somehow :(

**Comment by rkuhn** [comment]
@n1ko-w1ll That was what @drewhk was trying to say all along and what we have been very vocal about: you are testing a pre-release version of the first iteration of a new HTTP stack that has not yet been optimized at all. Of course this will not be representative of the performance when it reaches production quality.

Concerning the parameters: you cannot possibly have tested successfully with 10000 clients with the MacOS X default limit of 256 open file descriptors per process and 12000 file descriptors in the whole system, independent of which HTTP stack is used, and to my knowledge none of the popular operating system kernels come preconfigured with limits that are suitable for your specific test case. Doing load tests and optimization will always require dedicated configuration of all aspects of the deployment.

Another note on my measurement: the point was to show that the refused connections are indeed due to the limited queue of incoming connection requests, I have not tuned the JVM settings at all—in fact I don’t even know how much memory it had available. Finding out how to do these things with Maven goes beyond what time I can spend on this issue today, perhaps you can repeat the measurement with the patched version to get an even comparison.

**Comment by n1ko-w1ll** [comment]
I'm not sure if it makes any difference, but as mentioned in my initial post, I'm running Windows 7, not MacOS X ;)

I know that all of this is not released yet. I just want to point out, that it works out of the box without any adjustments on operating system level or any special configuration with spring boot and I wondered why this is not possible with akka when I initially tried it.

It's the same for me, I don't think I have the time today to test this again. Maybe next week.

**Comment by hepin1989** [comment]
Maybe we could make a online chart which show the benchmark for release iterations.cause akka have websocket and http now,best for microservice and rest facade.

**Comment by rkuhn** [comment]
From this I conclude that on Windows these kernel settings are not needed, and all that was required was to switch the higher backlog parameter value in your test case.

**Comment by n1ko-w1ll** [comment]
@rkuhn I tested your code on Windows and still received the "Connection refused" errors :(

**Comment by rkuhn** [comment]
We’ll have to revisit this when we are in a position to actually perform real benchmarks and performance analyses, thanks for providing the test case.

**Comment by jrudolph** [comment]
> Johannes, it might be interesting how only some of the requests get “stuck”, there is a clear latency gap between almost everything going nice and fast and some requests seem to starve.

Hard to say if there's really a gap. gatling output shows only > 1200ms and the ones which timeout above 60s. So, without further analyzing the output it could either be that some requests never finish or that the actual response times are just the long-tail of a regular distribution.

Here's the [tcpdump capture (torrent)](http://torcache.net/torrent/AD7FDEA411AAA39EBC000CCBFD9687E70B6AF653.torrent) of a benchmarking run.

If you look at this file and select an HTTP request and open the HTTP section it should show an entry "Response in Frame", which contains the frame number of the response. In the right-click-menu you can make this number a column in the frame table. This way you can scroll through the frames to get a feeling of how responses are distributed.

From what I've seen it looks like this for me:
 * long running requests are only first ones on a connection, latter requests are all handled in less than 2 seconds
 * every connection has seen at least one response (some connections are half-closed by the client after 60 seconds but the response will still be delivered)
 * i.e. most requests are run over only a small fraction of connections while other connections just hang somewhere in the loop before the first response is sent out
 * it seems during ramp down at the end stalled connections manage to send out at least one response
 * in the end 56 % of all connections handle only one request, while 50% of all requests is handled by just 15% of all connections

Here's a chart which shows one dot per connection for the number of responses received over the connection versus the first packet of the connection (excel analysis of output of wireshark's "Conversation List"), so the 20 second domain corresponds to the ramp up time:

![requestsperstarttime](https://cloud.githubusercontent.com/assets/9868/7554523/273620aa-f72c-11e4-8abf-5795a6233150.png)

This would somewhat support the assumption that there's a battle between resources for initializing connections and handling actual requests. The question remains how exactly the connection establishment work is buried under all the incoming requests (e.g. it could be either unfairness, the task being stuck in some queue, or the work itself is so big, that given the number of other work going on concurrently it just stretches over a too long time to be ever finished under the current total work load).

**Comment by jrudolph** [comment]
Looking through stack traces I think I know what happens: in `Http.bindAndHandle` we use `Sink.foreach` to handle incoming connections. This means per-connection materialization isn't handled in parallel but is linearized, instead, and is as such disadvantaged against other work that makes use of the rest of the thread-pool. We may consider using `mapAsync(Unordered)` instead and make the parallelism factor configurable so that you can choose how much of your thread-pool you want to sacrifice to materialization. With the workload of this load test, increasing the number will probably reduce throughput in favor of being able to accept more connections.

**Comment by rkuhn** [comment]
Many thanks, Johannes! Another problem with the current Streams is that 2.3.10 suffers from an Actor starvation bug that will be fixed in 2.3.11 (most likely tomorrow), and which can well have such an effect, see #17341.

**Comment by jrudolph** [comment]
I created #17512 to fix the possible connection establishment parallelization issue. The most pressing general streams performance issue (and regression from former releases) was that `validate` was slow and called in many place. This has already been addressed as #17377.

**Comment by jrudolph** [comment]
I'm closing this for now.

**Comment by dnvriend** [comment]
Spray seems to have the same problem as akka-http; only I don't believe the problem lies with Spray nor Akka-HTTP because the os has a max number of concurrent connections it can handle, so reaching for 10.000 concurrent connections is a bit much for desktop systems. I guess, opening and closing a connection would solve the problem. My OSX system can handle 2560 connections (the standard configuration) so maybe Spring does things differently?

I have created a destilled version of the sample application of Niko Will that to also try out Spray and AkkaHTTP. see: https://github.com/dnvriend/akka-poc

**Comment by n1ko-w1ll** [comment]
@dnvriend so SpringBoot does fail and refuses connections in your test? Interessting. As you can see from my posted Gatling result above, in my tests there was not even one rejected connection attempt. And as stated, the above posted output is from our real code which does some parsing and querying on the database.

I just started learning scala, so I'm not sure if I understand all your code. But from my point of view, you're using normal response bodies in your Spring implementation. We're using DeferredResult for asynchronous handling of the requests. Second thing, I don't know Spray, maybe extending SprayMarshallers does something different here, too.

**Comment by swachter** [comment]
Should this issue be fixed by now? I experience the same problem using version 1.0-RC4. I increased the backlog parameter, but that seemed to have no impact.

**Comment by drewhk** [comment]
There was a regression in RC4 that we have now PRs for. This ticket is not about that, though.

**Comment by jrudolph** [comment]
@swachter What is the "same problem" exactly? That many connections work on your OS with spring but not with akka-http? Which kernel settings did you tweak?

**Comment by swachter** [comment]
I work under Windows 7 without any tweaks. I prepared a small github project that demonstrates that Spring-Http copes better with heavy load than Akka-Http. The Spring server handles all requests without errors whereas the Akka server refuses a lot of connection requests.

https://github.com/swachter/akka-vs-spring

**Comment by drewhk** [comment]
The "reset-by-peer" events might have nothing to do with load, see the ticket here: https://github.com/akka/akka/issues/17854

**Comment by swachter** [comment]
I executed my test with the just released 1.0 version. Unfortunately, the problem still exists, i.e. akka-http refuses lots of connections where spring-http does not.

Maybe someone can have a look on it? The test project is still available (see my comment some days ago).

**Comment by ktoso** [comment]
// cc @sirthias @jrudolph

**Comment by jrudolph** [comment]
The fix for this ticket was to introduce `akka.http.server.max-connections` which specifies how many connections are handled in parallel. This defaults to 1024. If you want to handle more connections than that you have to increase that value.

With value changed to 10000 I can get about 4000 RPS (from ~ 1000 without the setting) with akka-http vs. 7000 I get with spring.

Modulo the connection concurrency above comments still hold: connection establishment is still 
somewhat penalized in akka-stream, so you will get bad RPS numbers with short tests like yours that mostly stress connection establishment.

**Comment by swachter** [comment]
As suggested I increased max-connections to 10000. The results improved somewhat (cf. https://github.com/swachter/akka-vs-spring/blob/master/README.md) but is still far from optimal.

In another test I added a processing delay of 2 seconds before the response is returned. 

```
complete {
     val promise = Promise[String]()
     scheduler.scheduleOnce(Duration(2000l, TimeUnit.MILLISECONDS))(promise.success("Akka"))
     promise.future
}
```

This worsened the result significantly. Now about half of the connection attempts where refused.

I do not understand how introducing the delay could impact the result in such a negative way.

**Comment by jrudolph** [comment]
Does the same happen with spring? What would your expectation be?

I'm also not particularly confident that using gatling (esp. with a ramp up / ramp down phase) is suitable to do a max throughput test. AFAIU gatling is more suited to check how fast your implementation is under the given scenario.

**Comment by swachter** [comment]
I changed the Spring implementation to return a DeferredResult and completed these results after 2000 ms (cf. my test project). Again, Spring did not refuse any connections and the collected statistics are as expected:

```
================================================================================
---- Global Information --------------------------------------------------------
> request count                                      44120 (OK=44120  KO=0     )
> min response time                                   2000 (OK=2000   KO=-     )
> max response time                                   2605 (OK=2605   KO=-     )
> mean response time                                  2023 (OK=2023   KO=-     )
> std deviation                                         57 (OK=57     KO=-     )
> response time 50th percentile                       2007 (OK=2007   KO=-     )
> response time 75th percentile                       2014 (OK=2014   KO=-     )
> mean requests/sec                                1381.167 (OK=1381.167 KO=-     )
---- Response Time Distribution ------------------------------------------------
> t < 800 ms                                             0 (  0%)
> 800 ms < t < 1200 ms                                   0 (  0%)
> t > 1200 ms                                        44120 (100%)
> failed                                                 0 (  0%)
================================================================================```

**Comment by jrudolph** [comment]
Ah, you meant about refused connections? I didn't have any of these with the longer timeouts. Maybe it's specific for Windows?

Here is how the akka-http version looks for me on Linux:

```
================================================================================
---- Global Information --------------------------------------------------------
> request count                                      34824 (OK=34824  KO=0     )
> min response time                                   2011 (OK=2011   KO=-     )
> max response time                                  17137 (OK=17137  KO=-     )
> mean response time                                  2857 (OK=2857   KO=-     )
> std deviation                                       1784 (OK=1784   KO=-     )
> response time 50th percentile                       2156 (OK=2156   KO=-     )
> response time 75th percentile                       3272 (OK=3272   KO=-     )
> mean requests/sec                                1073.754 (OK=1073.754 KO=-     )
---- Response Time Distribution ------------------------------------------------
> t < 800 ms                                             0 (  0%)
> 800 ms < t < 1200 ms                                   0 (  0%)
> t > 1200 ms                                        34824 (100%)
> failed                                                 0 (  0%)
================================================================================
```

**Comment by swachter** [comment]
Ok. I will repeat my tests under Linux. (The production environment is Linux anyway.)

What bothers me is that introducing a processing delay caused more connection refusals than before. I had expected that the non-blocking processing delay would remove some load and therefore would result in fewer connection refusals.

I am not a performance expert, in particular not on such intricate things like socket handling. Yet, I think that the Gatling test in combination with the comparison to the results returned by the Spring implementation may indicate a problem in the current connection handling.

I like Akka Stream and Akka Http. In particular its concise routing DSL. Therefore I hope that it is going to match the performance of the Spring implementation.

**Comment by swachter** [comment]
I repeated the test on a somewhat older Ubuntu box. After increasing ulimit to 20.000 I got no SocketExceptions neither with the Spring nor with the Akka version.

Here are the measured results:

Spring:

```
================================================================================
---- Global Information --------------------------------------------------------
> request count                                      39124 (OK=39124  KO=0     )
> min response time                                   2000 (OK=2000   KO=-     )
> max response time                                  54611 (OK=54611  KO=-     )
> mean response time                                  2407 (OK=2407   KO=-     )
> std deviation                                        862 (OK=862    KO=-     )
> response time 50th percentile                       2237 (OK=2237   KO=-     )
> response time 75th percentile                       2581 (OK=2581   KO=-     )
> mean requests/sec                                603.803 (OK=603.803 KO=-     )
---- Response Time Distribution ------------------------------------------------
> t < 800 ms                                             0 (  0%)
> 800 ms < t < 1200 ms                                   0 (  0%)
> t > 1200 ms                                        39124 (100%)
> failed                                                 0 (  0%)
================================================================================
```

Akka:

```
================================================================================
---- Global Information --------------------------------------------------------
> request count                                      28554 (OK=28554  KO=0     )
> min response time                                   2010 (OK=2010   KO=-     )
> max response time                                  53754 (OK=53754  KO=-     )
> mean response time                                  3935 (OK=3935   KO=-     )
> std deviation                                       5900 (OK=5900   KO=-     )
> response time 50th percentile                       2276 (OK=2276   KO=-     )
> response time 75th percentile                       3515 (OK=3515   KO=-     )
> mean requests/sec                                501.784 (OK=501.784 KO=-     )
---- Response Time Distribution ------------------------------------------------
> t < 800 ms                                             0 (  0%)
> 800 ms < t < 1200 ms                                   0 (  0%)
> t > 1200 ms                                        28554 (100%)
> failed                                                 0 (  0%)
================================================================================
```

**Comment by jrudolph** [comment]
Thanks, @swachter for checking.

@drewhk (when you are back): any idea how to proceed?

**Comment by drewhk** [comment]
@jrudolph what do you mean? Performance wise, it is the optimization of materialization that would matter most for HTTP right now. Or is there another issue here?

**Comment by jrudolph** [comment]
It seems the issue here (delaying responses will lead to connection refusals) is a Windows-only one. I don't have access to a Windows machine so I cannot really help with debugging.

**Comment by drewhk** [comment]
@swachter we had basically zero performance optimizations in place so far. This is one of the main post 1.0 goals, while 1.0 was to reach stable API/semantics. 

Anyway, I used  Windows to test Akka HTTP a few weeks ago because we debugged a regression and had ~3000 req/s with no connection refusals. I can do another test with delayed response.

**Comment by drewhk** [comment]
I reopened this, and will look into the delayes response scenario on Windows.

**Comment by rkuhn** [comment]
can this be closed?

**Comment by swachter** [comment]
The results are still far from optimal (at least on Windows 8); tested with Akka 2.4.2. I will repeat the test on Ubuntu soon.

Akka (on Windows 8):

```
================================================================================
---- Global Information --------------------------------------------------------
> request count                                      35564 (OK=18131  KO=17433 )
> min response time                                      1 (OK=2001   KO=1     )
> max response time                                   4462 (OK=4462   KO=1193  )
> mean response time                                  1557 (OK=2376   KO=707   )
> std deviation                                        992 (OK=575    KO=495   )
> response time 50th percentile                       2007 (OK=2022   KO=1024  )
> response time 75th percentile                       2022 (OK=2566   KO=1070  )
> mean requests/sec                                1153.776 (OK=588.21 KO=565.566)
---- Response Time Distribution ------------------------------------------------
> t < 800 ms                                             0 (  0%)
> 800 ms < t < 1200 ms                                   0 (  0%)
> t > 1200 ms                                        18131 ( 51%)
> failed                                             17433 ( 49%)
---- Errors --------------------------------------------------------------------
> java.net.ConnectException: Connection refused: no further info  11682 (67,01%)
rmation: localhost/127.0.0.1:9090
> java.net.ConnectException: No buffer space available (maximum    5751 (32,99%)
connections reached?): connect
================================================================================
```

In case someone want to try it - the test project is available at https://github.com/swachter/akka-vs-spring

**Comment by drewhk** [comment]
with persistent or not-persistent connections?

**Comment by swachter** [comment]
I repeated the test on my Linux box with Akka 2.4.2. The request count increased by ~ 25% compared to the test of July 21th, 2015 (1.0-experimental; see above). 

```
================================================================================
---- Global Information --------------------------------------------------------
> request count                                      34339 (OK=34339  KO=0     )
> min response time                                   2007 (OK=2007   KO=-     )
> max response time                                  40082 (OK=40082  KO=-     )
> mean response time                                  2851 (OK=2851   KO=-     )
> std deviation                                       2756 (OK=2756   KO=-     )
> response time 50th percentile                       2021 (OK=2021   KO=-     )
> response time 75th percentile                       2303 (OK=2303   KO=-     )
> mean requests/sec                                756.449 (OK=756.449 KO=-     )
---- Response Time Distribution ------------------------------------------------
> t < 800 ms                                             0 (  0%)
> 800 ms < t < 1200 ms                                   0 (  0%)
> t > 1200 ms                                        34339 (100%)
> failed                                                 0 (  0%)
================================================================================
```

@drewhk : AFAIK Gatling uses non-persistent connections. How can I check?

**Comment by drewhk** [comment]
We are aware of the performance bottleneck for non-persistent connections, it is the per-flow fusing that costs this much. We hadn't had the time to fully get around this (pre-fusing is possible with some trickery), but persistent connections are a completely different matter. If you put Akka HTTP behind a balancer that does the keep-alive for you you will see a difference even with gatling. I hope I will get the time to look at the non-persistent case again, here is my last attempt: https://github.com/akka/akka/pull/19762
As you see it is not much change that is needed but that attempt had some bugs and we were under pressure to release. I hope to get back to this soon.

**Comment by ktoso** [comment]
Closing as obsolete, we did a lot of work in the area recently, reopen if needed please

**Comment by swachter** [comment]
Just for the recored (ticket can stay closed): I repeated the test with the current Akka version (2.4.10) on the same Linux box I used on March 25th.  Again the performance increased significantly (about ~30 %):

```
================================================================================
---- Global Information --------------------------------------------------------
> request count                                      44114 (OK=44114  KO=0     )
> min response time                                   2009 (OK=2009   KO=-     )
> max response time                                   2223 (OK=2223   KO=-     )
> mean response time                                  2018 (OK=2018   KO=-     )
> std deviation                                         14 (OK=14     KO=-     )
> response time 50th percentile                       2011 (OK=2012   KO=-     )
> response time 75th percentile                       2021 (OK=2021   KO=-     )
> mean requests/sec                                1379.856 (OK=1379.856 KO=-     )
---- Response Time Distribution ------------------------------------------------
> t < 800 ms                                             0 (  0%)
> 800 ms < t < 1200 ms                                   0 (  0%)
> t > 1200 ms                                        44114 (100%)
> failed                                                 0 (  0%)
================================================================================
```

This result is even better as the result using the Spring version:

```
================================================================================
---- Global Information --------------------------------------------------------
> request count                                      41625 (OK=41625  KO=0     )
> min response time                                   2000 (OK=2000   KO=-     )
> max response time                                   3694 (OK=3694   KO=-     )
> mean response time                                  2173 (OK=2173   KO=-     )
> std deviation                                        271 (OK=271    KO=-     )
> response time 50th percentile                       2059 (OK=2059   KO=-     )
> response time 75th percentile                       2221 (OK=2221   KO=-     )
> mean requests/sec                                1304.573 (OK=1304.573 KO=-     )
---- Response Time Distribution ------------------------------------------------
> t < 800 ms                                             0 (  0%)
> 800 ms < t < 1200 ms                                   0 (  0%)
> t > 1200 ms                                        41625 (100%)
> failed                                                 0 (  0%)
================================================================================
```

TODO repeat the runs on a Windows box.
